test/cpp/end2end/xds/xds_end2end_test.cc

```diff
@@ -521,6 +521,19 @@ grpc_millis NowFromCycleCounter() {
   return grpc_timespec_to_millis_round_down(gpr_now(GPR_CLOCK_MONOTONIC));
 }
 
+// There can be milliseconds of difference between cycle time and time spec,
+// when the difference is in the order of a few seconds, we need to know that so
+// that our tests can account for the difference and reduce flakes due to the
+// clock difference.
+grpc_millis ClockDifference() {
```

Comment:
I think we probably want a more convenient interface here, so that the adjustment logic is all here instead of in the individual tests.  Consider something like the following:

```
MATCHER_P2(AdjustedClockInRange, t1, t2) {
  grpc_millis now = arg;
  // ...apply whatever adjustment to now is appropriate given the clock difference...
  bool ok = true;
  ok &= ::testing::ExplainMatchResult(
      ::testing::Ge(t1), arg, result_listener);
  ok &= ::testing::ExplainMatchResult(
      ::testing::Lt(t2), arg, result_listener);
  return ok;
}
```

Then, in the individual test, instead of doing this:

```
grpc_millis t1 = t0 + kTimeoutGrpcTimeoutHeaderMaxSecond * 1000 +
                 kTimeoutMillis - clock_difference;
grpc_millis t2 = t0 + kTimeoutMaxStreamDurationSecond * 1000 +
                 kTimeoutMillis + clock_difference;
// ...
t0 = NowFromCycleCounter();
EXPECT_GE(t0, t1);
EXPECT_LT(t0, t2);
```

You can instead do something like this:

```
grpc_millis t1 = t0 + kTimeoutGrpcTimeoutHeaderMaxSecond * 1000 +
                 kTimeoutMillis;
grpc_millis t2 = t0 + kTimeoutMaxStreamDurationSecond * 1000 +
                 kTimeoutMillis;
// ...
EXPECT_THAT(NowFromCycleCounter(), AdjustedClockInRange(t1, t2));
```

Reply:
done